### ISSUE TITLE
[Monitoring V1, RKE2] `422 Unprocessable entry` when installing monitoring v1, `Failed to ensure prometheus operator app: stale` on RKE2 clusters

### ISSUE THREAD

**Issue by slickwarren** [rancher/rancher#32969]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
* Deploy downstream rke2 cluster
* install monitoring v1 0.2.3 on this cluster

**Result:**
 `422 Unprocessable entry` error. Rancher also shows this error, as if it can't check that prometheus-operator isn't already installed: 
```
Failed to ensure prometheus operator app: stale "monitoring-operator" App in p-r58z5 Project is still on terminating
```

<img width="1510" alt="Screen Shot 2021-06-01 at 9 06 28 AM" src="https://user-images.githubusercontent.com/16691014/120355979-af8daa00-c2b8-11eb-9ad8-b50513277fe0.png">

**Other details that may be helpful:**
Once user tries to install monitoring and gets this error, the UI option to install monitoring goes away
<img width="398" alt="Screen Shot 2021-06-01 at 9 30 27 AM" src="https://user-images.githubusercontent.com/16691014/120359488-94249e00-c2bc-11eb-9c52-72916d927759.png">

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): master-head (5d49b92f479a00773607fb41fe851f98271148c0)
- Installation option (single install/HA): single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Digital Ocean 
- Machine type (cloud/VM/metal) and specifications (CPU/memory): 2vcpu, 4 worker nodes, 1 cp, 1 etcd
- Kubernetes version (use `kubectl version`): v1.21.1-alpha5+rke2r1

**Comment by aiyengar2** [comment]
Could not reproduce in a fresh RKE2 cluster based on master-head `4d1840056b02e390c29a9497abffc840c1de3b28`

Cluster:
- k8s version: `v1.21.1-rc3+rke2r1`
- 1 worker only, 1 all roles, both 8GB memory / 4CPU compute / 160GB disk

**Comment by slickwarren** [comment]
hmm, still seeing this on Rancher | master-head (d44698df1acb274d0ef55e8062271fb0dcf64ae3)
tried it with a single node (6vCPU, all roles) and the same setup as I had in the original (2vCPU, 3 workers, 1all roles), both result in `Uncaught (in promise) Failed while: Wait for action=viewMonitoring`. after trying a reinstall , I then get the `422 (Unprocessable Entity)` error.  Now, the `monitoring` option in the `tools` dropdown does persist.

**Comment by aiyengar2** [comment]
Was able to reproduce this on a RKE2 cluster but **not** on an RKE1 cluster. Not sure if this is one issue or multiple issues, so posting my investigation below.

The first time I tried to enable Monitoring in an RKE2 cluster, nothing happened and the UI was stuck on loading. However, I did get a log that `2021/07/06 22:43:05 [INFO] Redeploy Rancher Agents is needed for c-m-fef93c30: forceDeploy=false, agent/auth image changed=false, private repo changed=false, agent features changed=true`. 

The second time I tried to enable Monitoring (in the stuck loading page), the Helm deployment successfully went out. I was able to see Prometheus and Grafana provisioned in the cluster; however the Monitoring Pane itself did not show any update; it still showed the `Enable` button; therefore, disabling was not possible unless I manually cleaned up the resources

The Cluster object in the RKE2 cluster and the cattle-cluster-agent were also not updated, as seen below:

![Screen Shot 2021-07-06 at 4 06 01 PM](https://user-images.githubusercontent.com/13127443/124825610-b8f9da00-df28-11eb-86e8-2f3b0a1b4302.png)
![Screen Shot 2021-07-06 at 4 06 42 PM](https://user-images.githubusercontent.com/13127443/124825615-bac39d80-df28-11eb-9687-ebcac714db1f.png)

The third time I tried to enable Monitoring (after it was already deployed), I was able to reproduce the error:

![Screen Shot 2021-07-07 at 1 32 35 PM](https://user-images.githubusercontent.com/13127443/124825069-032e8b80-df28-11eb-9591-d8e1f39a29f7.png)

![Screen Shot 2021-07-07 at 1 32 42 PM](https://user-images.githubusercontent.com/13127443/124825828-f9595800-df28-11eb-9ca0-529c2e99b7f7.png)

![Screen Shot 2021-07-07 at 1 33 07 PM](https://user-images.githubusercontent.com/13127443/124825848-ffe7cf80-df28-11eb-9b8b-84641f55323f.png)

Since it was the `viewMonitoring` action that failed, I took a look at the cluster object and found the issue.

#### TLDR; Actions are missing on v3 CAPI based clusters

Below is a screenshot of RKE2 v.s. RKE1 Cluster objects:

![Screen Shot 2021-07-07 at 1 30 49 PM](https://user-images.githubusercontent.com/13127443/124826065-44736b00-df29-11eb-82de-9b6652f8e66d.png)
 
![Screen Shot 2021-07-07 at 1 30 40 PM](https://user-images.githubusercontent.com/13127443/124826082-49381f00-df29-11eb-8e28-272ad5fe2fc7.png)

In addition to `viewMonitoring` and `disableMonitoring`, which seems to be the root cause of this ticket, numerous other actions are also not available, which could be resulting in several issues across different legacy features.

**Comment by aiyengar2** [comment]
Based on offline conversation, the two paths forward here are:
1. To remove the ability to `enableMonitoring` for CAPI-provisioned clusters (e.g. RKE2 clusters or imported clusters created by the default Dashboard UI) and file a UI ticket to block showing the Monitoring V1 tab for clusters that do not support the `viewMonitoring` action (e.g. return this 422 message)
2. To add the ability to `disableMonitoring` and `viewMonitoring` for CAPI-provisioned clusters.  
 
For either action, this will likely involve debugging why only `enableMonitoring` is shown, which I suspect has to do with trying to update the shadow clusters.management.cattle.io/v3 object [when enabling Monitoring](https://github.com/rancher/rancher/blob/5aa95f2361df27376acd517637a8abfe08101dcb/pkg/api/norman/customization/cluster/actions_monitoring.go#L120-L124). It's possible that this either fails or that it gets overwritten on a sync of the original clusters.provisioning.cattle.io/v1 object that represents the CAPI-provisioned cluster. 

This would explain why we still see the Helm deployment occur (since `enableMonitoring` returns successfully) but we do not see any of the other actions come up on the menu since [they are shown only when `enableClusterMonitoring` is true](https://github.com/rancher/rancher/blob/5aa95f2361df27376acd517637a8abfe08101dcb/pkg/api/norman/customization/cluster/formatter.go#L92-L109).

Will update this ticket once an offline decision has been made of which route should be taken.

**Comment by aiyengar2** [comment]
On looking into this issue further, it seems like the issue definitely has to do with the provisioningv2 cluster sync controllers here:

https://github.com/rancher/rancher/blob/5aa95f2361df27376acd517637a8abfe08101dcb/pkg/controllers/provisioningv2/cluster/controller.go#L95-L114

The root cause of this issue is described in detail in the PR description in https://github.com/rancher/rancher/pull/35394.

---

To provide some context, in Rancher 2.6 we have two different cluster types:
- `v3` clusters: the legacy cluster type associated with the `rancher/norman` API and the legacy Cluster Manager UI, e.g.
  - RKE1 node driver clusters
  - custom clusters
  - clusters that have been imported via the Cluster Manager UI
- `v1` clusters: the new cluster type introduced in Rancher 2.6 that is associated with `rancher/steve` API and the new Cluster Dashboard UI (default in 2.6), e.g.
  - RKE2 node driver clusters
  - clusters that have been imported via the Cluster Dashboard UI.

In general, the Cluster Manager UI lists only `v3` clusters and the Cluster Dashboard UI lists only `v1` clusters, so to support both UIs showing both types of clusters, we introduced a set of cluster sync controllers that treat one of the clusters as the "master" cluster and automatically uses a feature within `rancher/wrangler` called `apply` (a Rancher precursor to [server-side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply)) to maintain a "shadow" cluster of the other type. e.g. if we create a `v1` cluster first, a shadow `v3` cluster will be created and if we create a `v3` cluster first, a shadow `v1` cluster will be created.

The problem with this approach is that **an update to a "shadow" `v3` cluster, such as enabling Monitoring, that is not tracked in a `v1` cluster field, will get immediately overridden**. This happens because, on [watching a change to a `v3` cluster](https://github.com/rancher/rancher/blob/5aa95f2361df27376acd517637a8abfe08101dcb/pkg/controllers/provisioningv2/cluster/controller.go#L108-L109), we [re-enqueue `v1` clusters](https://github.com/rancher/rancher/blob/5aa95f2361df27376acd517637a8abfe08101dcb/pkg/controllers/provisioningv2/cluster/controller.go#L129), which then causes us to [regenerate the cluster again from scratch based on the existing `v1` cluster](https://github.com/rancher/rancher/blob/5aa95f2361df27376acd517637a8abfe08101dcb/pkg/controllers/provisioningv2/cluster/controller.go#L95-L104). But since the `v1` cluster does not have a field for enabling a legacy feature, this synchronization reverts the change made to the shadow `v3` cluster.

We were able to confirm this by using https://github.com/ibuildthecloud/wtfk8s, which showed the cluster immediately flip back to `enableClusterMonitoring: false` on trying to enable Monitoring.

A fix that was attempted in https://github.com/rancher/rancher/pull/35394 was to check our caches on trying to create a new `v3` cluster based on an existing `v1` cluster to see if we can preserve the fields that do not need to be overridden rather than recreating it. However, that doesn't seem to have fixed the problem.
